### PR TITLE
Support for breakpoint maps (objects)?

### PR DESCRIPTION
**Pull request by WebSeed**
Hi,

I like the terseness of current responsive array syntax but in some situations I think it would be nice to provide breakpoint maps as well as arrays. For example, where you have a lot of breakpoints to avoid so many `null` values, or where you may need to insert new breakpoints.

Example:

```jsx
<Box p={[1, null, 2, null, null, 3]}>
```

Becomes:

```jsx
<Box p={{ def: 1, sm: 2, xl: 3 }}>
```

Where `theme.breakpoints` is no longer an array but something like `{ def: 0, xs: '32em', sm: '40em', md: '56em', lg: '64em', xl: '80em' }`.

That's what this PR introduces, in what is (at least according to the tests) a non-breaking way, as well as moving some duplicated code to shared function `getStyles`.

I figure you may well have considered and dismissed it from an API design perspective. But I'm going to use it in the fork, so just thought I'd share back 😄 . 

I can remove the breakpoint map changes if you are interested in some of the lightweight refactoring / removal of duplication between the `style` and `space` functions.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -8,8 +8,8 @@
   "scripts": {
     "prepare": "npm run clean && npm run build:cjs && npm run build:esm",
     "clean": "rm -rf dist && mkdir dist",
-    "build:cjs": "cross-env NODE_ENV=cjs babel src -o dist/index.cjs.js",
-    "build:esm": "cross-env NODE_ENV=esm babel src -o dist/index.esm.js",
+    "build:cjs": "cross-env NODE_ENV=cjs babel src -o dist/index.cjs.js --source-maps",
```
[the remark this answers is not in the document]
Comment (WebSeed) [reply package.json]: Yes, you're right. Removed.

File: src/index.js [hunk 2]
```diff
@@ -19,6 +19,8 @@ export const defaultBreakpoints = [ 40, 52, 64, ].map(n => n + 'em')
 export const is = n => n !== undefined && n !== null
 export const num = n => typeof n === 'number' && !isNaN(n)
 export const px = n => num(n) ? n + 'px' : n
+export const isArray = n => Array.isArray(n)
```
[the remark this answers is not in the document]
Comment (WebSeed) [reply src/index.js]: Must admit, somehow I didn't think of just destructuring the fn from `Array` - I've put that in.

**Comment by tbranyen** [comment]
I really like this feature. The current approach is more terse, but our theme currently uses the small, medium, large nomenclature, so this would be a welcome addition for consistency reasons.

**Comment by Millsky** [comment]
@WebSeed  - I was just thinking about putting in a issue to add an optional object mapping this morning, if you want an additional PR to update the documentation let me know, I can help out.

**Comment by WebSeed** [comment]
Hi @Millsky - thanks. Was thinking of putting some docs in if @jxnblk expressed interest in the functionality. But happy to collaborate if it gets the go ahead.

**Comment by jxnblk** [comment]
Sorry, I’ve been traveling the past few weeks. Planning on taking a closer look over the next few days

File: src/index.js [hunk 3]
```diff
@@ -50,6 +52,39 @@ export const compose = (...funcs) => {
 
 export const createMediaQuery = n => `@media screen and (min-width: ${px(n)})`
 
+const getStyles = (props, style, val) => {
```
Comment (jxnblk) [inline src/index.js]: For some reason, it feels like pulling this out from the other function makes it a little harder to compare with the previous implementation and don’t see much re-use in the `getSttyles` function – would you mind putting this back in the main function body?

Comment (WebSeed) [reply src/index.js]: Sure no problem. Will result in a little bit of duplication across top-level `space` and `style`  fns (~30 LoC).

File: src/index.js [hunk 4]
```diff
@@ -50,6 +52,39 @@ export const compose = (...funcs) => {
...
+        const media = createMediaQuery(breakpoints[i - 1])
+        styles[media] = rule
+      }
+    }
+    return styles
+  }
+
+  if (isArray(breakpoints)) return null
+
+  const styles = style(val.def) || {}
```
[the remark this answers is not in the document]
Comment (WebSeed) [reply src/index.js]: Yeah, I think `def` is a bit of an overloaded abbreviation - I did wonder about something a bit more abstract e.g. `_` or perhaps `default`? Allowing for a `default` entry feels a bit more in keeping with having the first entry of the array also not have a MQ associated?

But if you think it's best to add explicit entry to `theme.breakpoints` then I guess we could make `null` value represent non-MQ entry? E.g.

```jsx
theme.breakpoints = { default: null, sm: '20em' }
// ...
<Box p={{ default: '10px', sm: '20px' }}>
```

Would output:
```css
.xyz { padding: 10px; }
@media screen and (min-width) {
  .xyz { padding: 20px; }
}
```

File: src/index.js [hunk 5]
```diff
@@ -189,38 +200,15 @@ export const space = props => {
 
   return keys
     .map(key => {
-      const value = props[key]
+      const val = props[key]
```
[the remark this answers is not in the document]
Comment (WebSeed) [reply src/index.js]: When I refactored some of the functionality into `getStyles` I noticed that `val` was used in `export const style` but `value` was used in `export const space` - felt like they represented the same thing. Happy to change it back though - especially if removing `getStyles` (it was name of arguments to `getStyles(props, style, val)` that I was trying to keep consistent too).

File: src/index.js [hunk 6]
```diff
@@ -50,6 +52,39 @@ export const compose = (...funcs) => {
...
+      const rule = style(val[i])
+      if (rule) {
+        const media = createMediaQuery(breakpoints[i - 1])
+        styles[media] = rule
+      }
+    }
+    return styles
+  }
+
+  if (isArray(breakpoints)) return null
```
[the remark this answers is not in the document]
Comment (WebSeed) [reply src/index.js]: Hi, yeah - I could remove this and then the following test would pass:

```js
test('width returns responsive values for object', t => {
  const a = width({width: {0: 1, 2: 1/2}})

  t.deepEqual(a, {
    '@media screen and (min-width: 40em)': {
      width: '100%'
    },
    '@media screen and (min-width: 64em)': {
      width: '50%'
    },
  })
})
```

Not sure if this is confusing though - would you expect `0` to be without MQ?

**Comment by WebSeed** [comment]
Hi @jxnblk, I've made a few changes following your review https://github.com/jxnblk/styled-system/pull/341/commits/68fea521919ad7d8e5201224a5d3d4d663655845

Will just wait confirmation on a few of the other points - in particular whether you're happier with duplication of `getStyles` type functionality - and I can make the necessary changes. I'm also going to be on a holiday for a few days, so might be next weekend before I can update.

**Comment by tbranyen** [comment]
@jxnblk word, no worries. Just wanted to let you know there was active interest in this feature. This adds another layer of consistency to the API we're exploring.

File: src/index.js (hunk 3 above)
Comment (jxnblk) [reply src/index.js]: Ah, I see now. Yeah, `space` is an odd-ball, and I'd like to try to get rid of that (in favor of using `style`) in the next major version – it's probably fine to leave it split out then, but could be nice to use a single object argument like the other functions do

File: src/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (jxnblk) [reply src/index.js]: That's probably sloppiness on my end. Would be nice to use the more explicit `value` where it makes sense

File: src/index.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (jxnblk) [reply src/index.js]: Yeah, I would assume that `width: { 0: 1 }` would reference `breakpoints[0]` – generally I think the object value approach will be slightly more confusing (but also more flexible) than the array values, but should be a more explicit API

**Comment by WebSeed** [comment]
@jxnblk I think it might be ready for review again - I've changed `val` to `value` and allow for `width={{0: 1}}` syntax. If you're happy with the functionality should we take a look at the docs? Looks to me like there could be a single example in the [root README](https://github.com/jxnblk/styled-system#responsive-style-props) and perhaps another section in [docs/responsive-style](https://github.com/jxnblk/styled-system/blob/master/docs/responsive-styles.md)?

**Comment by WebSeed** [comment]
Hi @jxnblk - I've taken a quick pass at adding some documentation for this feature. Could you take another look please?

BTW - I've noticed a very similar feature (system with breakpoint map support) is being developed for Material UI (https://deploy-preview-13632--material-ui.netlify.com/system/basics/#object). I did consider whether you could have best of both worlds e.g. separate out `breakpoints` and `breakpointsMap` to allow for both `array` and `object` syntax but thought it might be simpler to just configure one vs. the other?

**Comment by jxnblk** [comment]
Looks like the addition of prettier in #352 has introduced some conflicts – sorry about that 😬 – I think they should be fairly straight forward, and this PR LGTM. I can resolve the conflicts with the GitHub UI unless you wanna take a quick look first

**Comment by WebSeed** [comment]
Hi @jxnblk @StevenLangbroek 

No problem should be fixed now - much better with Prettier in place! Also [applied Prettier to the tests too](https://github.com/jxnblk/styled-system/pull/341/commits/bb523e5132013acf46a8a769418d78c4a81f8549)

Merge when you're happy @jxnblk 😄

**Pull request merged**